File: pluck/pipelines/chapters.py
"""Split a video by its chapters into separate files, delivered as a .zip."""
import re
from pathlib import Path

import yt_dlp

from ..ytdlp import ydl_base
from .base import JobCtx, download_source, run_ffmpeg, zip_files


def _safe(name: str) -> str:
    return re.sub(r"[^\w\- ]+", "_", name).strip()[:60] or "chapter"


def _get_chapters(url: str) -> list[dict]:
    with yt_dlp.YoutubeDL({**ydl_base(), "skip_download": True}) as ydl:
        info = ydl.extract_info(url, download=False)
    return info.get("chapters") or []


def split_chapters(src: Path, chapters: list[dict], out_dir: Path) -> list[Path]:
    parts: list[Path] = []
    ext = src.suffix
    for i, ch in enumerate(chapters):
        start, end = ch.get("start_time"), ch.get("end_time")
        title = _safe(ch.get("title") or f"chapter-{i + 1}")
        part = out_dir / f"{i + 1:02d}-{title}{ext}"
        seek = []
        if start is not None:
            seek += ["-ss", str(start)]
        if end is not None:
            seek += ["-to", str(end)]
        r = run_ffmpeg([*seek, "-i", str(src), "-c", "copy", str(part)], timeout=300)
        if r.returncode == 0 and part.exists() and part.stat().st_size > 0:
            parts.append(part)
    if not parts:
        raise RuntimeError("chapter split produced no files")
    return parts


def run(ctx: JobCtx) -> Path:
    url = ctx.req.url.strip()
    chapters = _get_chapters(url)
    if not chapters:
        raise RuntimeError("no chapters")
    src = download_source(ctx, height_cap=1080)
    ctx.update(status="processing")
    parts = split_chapters(src, chapters, ctx.job_dir)
    src.unlink(missing_ok=True)
    zip_path = ctx.job_dir / f"{src.stem}-chapters.zip"
    return zip_files(parts, zip_path)

File: node/src/pipelines/chapters.js
/** Split a video by its chapters into separate files, delivered as a .zip.
 * Mirrors pluck/pipelines/chapters.py. */
import path from "node:path";
import { statSync } from "node:fs";

import { dumpJson } from "../ytdlp.js";
import { downloadSource, runFfmpeg, unlinkQuiet, zipFiles } from "./base.js";

function safe(name) {
  return (name || "").replace(/[^\w\- ]+/g, "_").trim().slice(0, 60) || "chapter";
}

async function getChapters(url) {
  const info = await dumpJson(url, ["--no-playlist"]);
  return info.chapters || [];
}

function ok(p) {
  try { return statSync(p).size > 0; } catch { return false; }
}

export async function splitChapters(src, chapters, outDir) {
  const parts = [];
  const ext = path.extname(src);
  for (let i = 0; i < chapters.length; i++) {
    const ch = chapters[i];
    const start = ch.start_time, end = ch.end_time;
    const title = safe(ch.title || `chapter-${i + 1}`);
    const part = path.join(outDir, `${String(i + 1).padStart(2, "0")}-${title}${ext}`);
    const seek = [];
    if (start != null) seek.push("-ss", String(start));
    if (end != null) seek.push("-to", String(end));
    const r = await runFfmpeg([...seek, "-i", src, "-c", "copy", part], 300_000);
    if (r.returncode === 0 && ok(part)) parts.push(part);
  }
  if (!parts.length) throw new Error("chapter split produced no files");
  return parts;
}

export async function run(ctx) {
  const url = ctx.req.url.trim();
  const chapters = await getChapters(url);
  if (!chapters.length) throw new Error("no chapters");
  const src = await downloadSource(ctx, { heightCap: 1080 });
  ctx.update({ status: "processing" });
  const parts = await splitChapters(src, chapters, ctx.jobDir);
  const stem = path.basename(src, path.extname(src));
  unlinkQuiet(src);
  const zipPath = path.join(ctx.jobDir, `${stem}-chapters.zip`);
  return zipFiles(parts, zipPath);
}
